include/GoogleSync/GoogleSyncBase.php

```diff
@@ -0,0 +1,1049 @@
+            return false;
+        }
+        return $authJson_local;
+    }
+
+    /**
+     * Creates and Sets the Google client in the object
+     *
+     * @param string $id : the SuiteCRM user id
+     *
+     * @return bool Success/Failure
+     * @throws GoogleSyncException
+     */
+    protected function setClient($id)
+    {
+        $isValidator = new SuiteValidator();
+        if (!$isValidator->isValidId($id)) {
+            throw new GoogleSyncException('Google Sync Base trying to set Client but given an invalid ID: ' . $id, GoogleSyncException::INVALID_CLIENT_ID);
+        }
+        
+        if (!$gClient_local = $this->getClient($id)) {
+            return false;
+        }
+        $this->gClient = $gClient_local;
+        return true;
+    }
+
+    /**
+     * Set the Google client up for the user by id
+     *
+     * @param string $id : the SuiteCRM user id
+     *
+     * @return \Google_Client|false Google_Client on success. False on failure.
+     * @throws GoogleSyncException if unable to retrive the user
+     */
+    protected function getClient($id)
+    {
+        $isValidator = new SuiteValidator();
+        if (!$isValidator->isValidId($id)) {
+            throw new GoogleSyncException('GoogleSyncBase trying to get Client but given ID is invalid: ' . $id, GoogleSyncException::INVALID_CLIENT_ID);
+        }
+
+        // Retrieve user bean
+        if (!isset($this->workingUser)) {
+            $this->workingUser = BeanFactory::getBean('Users', $id);
+            if (!$this->workingUser) {
+                throw new GoogleSyncException('Unable to retrieve a User bean', GoogleSyncException::UNABLE_TO_RETRIEVE_USER);
+            }
+        }
+
+        // Retrieve Access Token JSON from user preference
+        $accessToken = json_decode(base64_decode($this->workingUser->getPreference('GoogleApiToken', 'GoogleSync')), true);
+        if (!array_key_exists('access_token', $accessToken)) {
+            // The Token is invalid JSON or missing
+            // TODO: It should be an exception (do not leave same 'false' return value for differenc failure reason, Inform the caller about this problem)
+            $this->logger->fatal(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Invalid or Missing AuthToken');
+            return false;
+        }
+        // The refresh token is only provided once, on first authentication. It must be added afterwards.
+        if (!array_key_exists('refresh_token', $accessToken)) {
+            $accessToken['refresh_token'] = base64_decode($this->workingUser->getPreference('GoogleApiRefreshToken', 'GoogleSync'));
+        }
+
+        // New Google Client and refresh the token if needed
+        try {
+            $client = $this->getGoogleClient($accessToken);
+        } catch (Exception $e) {
+            // TODO: handle the exception instead just loggin it, it will be logged if unhandled by suitecrm core.
```

Comment:
@gymad 
I want to make sure that the sync process can continue for other users if the current one fails. If I leave this uncaught, will it stop the scheduled job when thrown?

Reply:
Fixed